src/bls12_381/ec.rs

```diff
@@ -166,6 +143,38 @@ macro_rules! curve_impl {
             fn into_projective(&self) -> $projective {
                 (*self).into()
             }
+
+            fn hash(t: Self::Base) -> Self
+            {
+                use ::LegendreSymbol::*;
+
+                // w = (t^2 + b + 1)^(-1) * sqrt(-3) * t
+                let mut w = t;
+                w.square();
+                w.add_assign(&Self::get_coeff_b());
+                w.add_assign(&Self::Base::one());
+                w = w.inverse().unwrap();
+                w.mul_assign(&Self::get_c1());
+                w.mul_assign(&t);
+
+                let mut x = w;
+                for i in 0..3 {
+                    match i {
+                        0 =>  { x.mul_assign(&t); x.negate(); x.add_assign(&Self::get_c2()) },
+                        1 =>  {                   x.negate(); x.sub_assign(&Self::Base::one()) },
+                        2 =>  { x=w; x.square();  x = x.inverse().unwrap(); x.add_assign(&Self::Base::one()) },
+                        _ =>  {}
+                    }
+
+                    if let Some(mut y) = Self::y2_from_x(x).sqrt() {
+                        if let QNonResidue = t.legendre() { y.negate() }
+                        return Self {x: x, y: y, infinity: false}
+                    }
+
+                }
+                Self {x: Self::Base::zero(), y: Self::Base::zero(), infinity: true }
```

Comment:
That is just Rust not understanding that the function it will return before for sure. That line is never executed.

I wasn't really sure what would have been conventional to do there, whether `panic!` or return a silly point like infinity - bn.h returns an `ERR_POINT` as far as I remember.
Plus, from ocaml I had this feeling that exceptions are bad and that I should keep my functions pure, so …

In any case, I'm happy to change it if needed!
As a side note, making it constant-time will probably get rid of that for loop and the silly return
(as described on the paper). For now I am aiming at a state that allows me to generate test-vectors.